File: python_pkg/python_pkg/cmd_vel_pub.py
# !/usr/bin/env/ python3
from geometry_msgs.msg import Twist
import rclpy
from rclpy.node import Node
import os

class CmdVelPublisher(Node):

    def __init__(self):
        super().__init__('cmd_vel_pub_node')
        ROBOT_MODEL = os.environ['ROBOT_MODEL']
        self.publisher = self.create_publisher(
            Twist, ROBOT_MODEL+'/cmd_vel', 10
        )  # queue size
        timer_period = 0.5  # seconds
        self.timer = self.create_timer(timer_period, self.publishCallback)
        self.get_logger().info(
            'DriveForward node Started, move forward during 5 seconds \n'
        )

    def publishCallback(self):
        twist_msg = Twist()
        twist_msg.linear.x = 0.5
        twist_msg.angular.z = 1.0
        self.publisher.publish(twist_msg)

    def stopRobot(self):
        stop_msg = Twist()
        stop_msg.linear.x = 0.0
        stop_msg.angular.z = 0.0
        self.publisher.publish(stop_msg)

def main(args=None):
    rclpy.init(args=args)

    cmd_vel_publisher = CmdVelPublisher()
    start_time = cmd_vel_publisher.get_clock().now().to_msg().sec
    clock_now = start_time
    time_delta = 0

    while (clock_now - start_time) < 5:
        rclpy.spin_once(cmd_vel_publisher)
        clock_now = cmd_vel_publisher.get_clock().now().to_msg().sec

        time_delta = clock_now - start_time
        cmd_vel_publisher.get_logger().info(f'{time_delta} seconds passed')

    cmd_vel_publisher.stopRobot()

    cmd_vel_publisher.get_logger().info('\n==== Stop Publishing ====')
    cmd_vel_publisher.destroy_node()

    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: cxx_pkg/src/cmd_vel_pub.cpp
#include <memory>
#include "geometry_msgs/msg/twist.hpp"
#include "rclcpp/rclcpp.hpp"

class TwistPub : public rclcpp::Node
{
 public:
  TwistPub(const std::string& model);
  void move_robot()
  {
    m_twist_msg.linear.x = 0.5;
    m_twist_msg.angular.z = 1.0;
    m_pub->publish(m_twist_msg);
  }
  void stop_robot()
  {
    m_twist_msg.linear.x = 0.0;
    m_twist_msg.angular.z = 0.0;
    m_pub->publish(m_twist_msg);
    RCLCPP_INFO(get_logger(), "==== Stop Robot ====");
  }

 private:
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr m_pub;
  rclcpp::TimerBase::SharedPtr m_timer;
  geometry_msgs::msg::Twist m_twist_msg;
  void timer_callback()
  {
    move_robot();
  }
};

TwistPub::TwistPub(const std::string& model): Node("cmd_vel_pub_node")
{
  RCLCPP_INFO(get_logger(), "Cmd_vel Pub Node Created");
  m_pub = create_publisher<geometry_msgs::msg::Twist>(model + "/cmd_vel", 10);
  m_timer = create_wall_timer(std::chrono::milliseconds(100),
                              std::bind(&TwistPub::timer_callback, this));
}

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  const std::string& model = std::getenv("ROBOT_MODEL");
  auto twist_pub = std::make_shared<TwistPub>(model);
  auto t_start = twist_pub->now();
  auto t_now = twist_pub->now();
  auto stop_time = 5.0;
  while ((t_now - t_start).seconds() < stop_time) {
    t_now = twist_pub->now();
    twist_pub->move_robot();
    RCLCPP_INFO(twist_pub->get_logger(), "%f Seconds Passed", (t_now - t_start).seconds());
  }
  twist_pub->stop_robot();
  rclcpp::shutdown();

  return 0;
}